tests/core/model_management/test_local_model_trainer.py

```diff
@@ -0,0 +1,209 @@
+    """
+    trainer = local_trainer(**trainer_type_cfg)
+    model_future = trainer.train(
+        SampleModule,
+        training_data=DataStream.from_iterable([]),
+        save_path=save_path,
+        save_with_id=True,
+    )
+    model_future.wait()
+    assert model_future.save_path != save_path
+    assert model_future.id in model_future.save_path
+    assert os.path.exists(model_future.save_path)
+
+
+def test_cancel_clean_termination(trainer_type_cfg):
+    """Test that cancelling an in-progress training successfully destroys the
+    training when the training is run in a way that can be shut down cleanly
+    """
+    trainer = local_trainer(**trainer_type_cfg)
+    model_future = trainer.train(
+        SampleModule,
+        training_data=DataStream.from_iterable([]),
+        sleep_time=1000,
+    )
+    assert model_future.get_status() == ModelTrainerBase.TrainingStatus.RUNNING
+    assert not model_future.get_status().is_terminal
+
+    # Cancel the future
+    model_future.cancel()
+    assert model_future.get_status() == ModelTrainerBase.TrainingStatus.CANCELED
+    assert model_future.get_status().is_terminal
+    model_future.wait()
+
+
+def test_cancel_without_waiting(trainer_type_cfg):
+    """Test that cancelling an in-progress training that uses a long sleep (and
+    can't be easily destroyed in a thread) still reports CANCELED as the status
+    before the training has fully terminated.
+    """
+    trainer = local_trainer(**trainer_type_cfg)
+    model_future = trainer.train(
+        SampleModule,
+        training_data=DataStream.from_iterable([]),
+        sleep_time=0.5,
+        sleep_increment=0.5,
+    )
+    assert model_future.get_status() == ModelTrainerBase.TrainingStatus.RUNNING
+    assert not model_future.get_status().is_terminal
+
+    # Cancel the future and make sure it reports canceled, even though the
+    # function is still sleeping
+    model_future.cancel()
+    assert model_future.get_status() == ModelTrainerBase.TrainingStatus.CANCELED
+    assert model_future.get_status().is_terminal
+    model_future.wait()
+
+
+def test_no_retention_time(trainer_type_cfg):
+    """Test that constructing with no configured retention period keeps futures
+    forever and doesn't cause errors
+    """
+    trainer = local_trainer(retention_duration=None, **trainer_type_cfg)
+    model_future = trainer.train(SampleModule, DataStream.from_iterable([]))
+    model_future.wait()
+    retrieved_future = trainer.get_model_future(model_future.id)
+    assert retrieved_future is model_future
+
+
+def test_purge_retention_time(trainer_type_cfg):
+    """Test that purging old models works as expected"""
+    trainer = local_trainer(retention_duration="0.1s", **trainer_type_cfg)
+    model_future = trainer.train(SampleModule, DataStream.from_iterable([]))
+    model_future.wait()
+    retrieved_future = trainer.get_model_future(model_future.id)
+    assert retrieved_future is model_future
+    time.sleep(0.2)
```

Comment:
oof, there's no way that this doesn't end up flaking out at some point though.
Is it possible to change a timestamp here instead to make this thing purge?
e.g.
```
    model_future = trainer.train(SampleModule, DataStream.from_iterable([]))
    model_future.wait()
    retrieved_future = trainer.get_model_future(model_future.id)
    assert retrieved_future is model_future
    model_future._completion_time = now() - 1 day
    with pytest.raises(ValueError):
        trainer.get_model_future(model_future.id)
```

Reply:
Yep, much cleaner